File: app/src/main/java/com/anthonymandra/rawdroid/data/AppDatabase.java
package com.anthonymandra.rawdroid.data;

import android.arch.persistence.room.Database;
import android.arch.persistence.room.Room;
import android.arch.persistence.room.RoomDatabase;
import android.content.Context;

@Database(entities = {
		FolderEntity.class,
		MetadataEntity.class,
		SubjectEntity.class,
		SubjectJunction.class,
		SynonymEntity.class},
		version = 1)
public abstract class AppDatabase extends RoomDatabase {
	private static final String DB_NAME = "curator.db";
	private static AppDatabase sInstance;

	public static AppDatabase getInstance(final Context context) {
		if (sInstance == null) {
			synchronized (AppDatabase.class) {
				if (sInstance == null) {
					sInstance = create(context.getApplicationContext(), false);
				}
			}
		}
		return sInstance;
	}

	static AppDatabase create(Context ctxt, boolean memoryOnly) {
		RoomDatabase.Builder<AppDatabase> b;

		if (memoryOnly)	//FIXME: Remove this test-only code
		{
			b = Room.inMemoryDatabaseBuilder(ctxt.getApplicationContext(), AppDatabase.class);
		}
		else
		{
			b = Room.databaseBuilder(ctxt.getApplicationContext(), AppDatabase.class, DB_NAME);
		}
		return(b.build());
	}

	public abstract FolderDao folderDao();
	public abstract MetadataDao metadataDao();
	public abstract SubjectDao subjectDao();
	public abstract SubjectJunctionDao subjectJunctionDao();
}

File: app/src/main/java/com/anthonymandra/rawdroid/data/AppDatabase.kt
package com.anthonymandra.rawdroid.data

import androidx.room.Database
import androidx.room.Room
import androidx.room.RoomDatabase
import android.content.Context
import androidx.room.migration.Migration
import androidx.sqlite.db.SupportSQLiteDatabase

@Database(entities = [
    FolderEntity::class,
    MetadataEntity::class,
    RecycleBinEntity::class,
    SubjectEntity::class,
    SubjectJunction::class,
    SynonymEntity::class],
    version = 3)
abstract class AppDatabase : RoomDatabase() {

    abstract fun folderDao(): FolderDao
    abstract fun metadataDao(): MetadataDao
    abstract fun recycleBinDao(): RecycleBinDao
    abstract fun subjectDao(): SubjectDao
    abstract fun subjectJunctionDao(): SubjectJunctionDao

    companion object {
        private const val DB_NAME = "curator.db"
        @Volatile private var INSTANCE: AppDatabase? = null

        fun getInstance(context: Context): AppDatabase =
            INSTANCE ?: synchronized(this) {
					INSTANCE ?:
						Room.databaseBuilder(context.applicationContext, AppDatabase::class.java, DB_NAME)
							.addMigrations(MIGRATION_2_3)
							.build()
							.also { INSTANCE = it }
            }

		 val MIGRATION_2_3 = object : Migration(2, 3) {
			 override fun migrate(database: SupportSQLiteDatabase) {
				 database.execSQL("CREATE TABLE IF NOT EXISTS `recycle_bin` (`id` INTEGER PRIMARY KEY AUTOINCREMENT NOT NULL, `path` TEXT NOT NULL)")
			 }
		 }
    }
}